### DISCUSSION TITLE
Github action to checkout repository

### DISCUSSION THREAD

**Discussion by somehowchris** [GitoxideLabs/gitoxide#401] [Ideas]
First, let me start with, damn that project is awesome 😄 

I've dug through it the last few hours as its really interesting but I find it a bit hard to get a grip where to start, as it seems to me its really different then using something like `git2-rs` (don't get me wrong, I dont like `git2-rs` but its somewhat close to what I use with git in a cli)

Before using it in something like `cargo-binstall` I would like to have a go with something else to learn and see how it works.

I had a search for something like a github action using `gitoxide` to replace `actions/checkout` but couldn't find one. It seems to me like this would ge a nice case to get the word out about `gitoxide` and aswell interms of efficenty (human time and hard ressources of ci).

May I ask, is it too early for something like this? If not, may I ask for some guidance on starting points for the usage of `gitoxide` or its subsequent crates as libraries?

Really appriciate the effort and any guidance 🦄

**Comment by Byron** [comment]
Thanks a lot for your interest! Indeed it's a bit tough to find your bearings within the 20+ crates or so, but `git-repository` is the top-level for application developers akin to `git2`. I'd build the docs (`cargo doc --open --package git-repository`) and go from `open(…)` or `discover(…)`.

Using `gitoxide` to accelerate checkouts on `CI` is an excellent idea, and I have added it to the [tracking issue](https://github.com/Byron/gitoxide/issues/303) that you could subscribe to to stay informed about the checkout progress.

At the current time, it's indeed a little early to use `gitoxide` for cloning and checkouts, but it's a good time if you want to read data from a git repository.

**Reply by somehowchris** [reply]
Really appriciate the guidance 💯 

Using `gitoxide` for a checkout action was just an idea from me to get some experience using it 😓 

I think a lot of crates such as `crates-index` (https://github.com/frewsxcv/rust-crates-index/issues/82) could also benefit as the crates git index is growing fast

**Reply by somehowchris** [reply]
BTW if `gitoxide` could be built as a static binary, the github action could just download a docker image built from scratch which means it's really just loading some mb and not a whole runtime

**Comment by Byron** [comment]
> I think a lot of crates such as crates-index (https://github.com/frewsxcv/rust-crates-index/issues/82) could also benefit as the crates git index is growing fast.

Indeed, it's already on [the list of crates](https://github.com/Byron/gitoxide/projects/1#column-14882365) that will be 'oxidized' one day.

> BTW if gitoxide could be built as a static binary [...]

That should be a decent fallback in case there is no binary support for a certain CI platform, even though it should be possible to build `gitoxide` for all platforms currently supported by GitHub.

**Reply by somehowchris** [reply]
Nice 👍🏻 

This "BTW" was more meant like, if speed matters, and the setup should be fairly easy without needing rust before the checkout `scratch` docker images could be a minimalistic thing

**Reply by Byron** [reply]
> This "BTW" was more meant like, if speed matters, and the setup should be fairly easy without needing rust before the checkout

Agreed! Building `gitoxide` before checking anything out would certainly negate any benefit. The idea is to download pre-built _binaries_ instead. If there is no such _binary_ for the platform at hand but docker is available, one could leverage this by running a docker image instead.

**Reply by somehowchris** [reply]
Github actions can be built as docker container actions, https://docs.github.com/en/actions/creating-actions/creating-a-docker-container-action so the version of a docker image could resemble the version of the action aka the version of the code behind the docker image

**Comment by Byron** [comment]
Thanks for the heads-up! I wasn't aware docker-actions have their own section in the manual and made sure I won't forget about that in the [tracking issue](https://github.com/Byron/gitoxide/issues/303). This should help to get something going quickly.
